### PR TITLE
Refactor handler

### PR DESCRIPTION
**Pull request by aereal**
Introduced these classes:

* Intercepter
  * manages registering/unregistering Handlers
  * try to intercept the request if matched
* Handler
  * works like a Plack application
  * provides a factory method that creates a instance from:
    * pattern-like: string or regular expression
    * app-like: a subroutine or an array reference that satisfies PSGI

### CODE REVIEW

**Comment by aereal** [comment]
@astj Please review this :pray:

**Comment by astj** [comment]
Please add `Test::WWW::Stub::Handler` and `Test::WWW::Stub::Intercepter` to `t/00_compile.t` so that we can test individual modules has no syntax errors.

File: lib/Test/WWW/Stub.pm [hunk 1]
```diff
@@ -59,19 +48,14 @@ sub import {
 sub register {
     my ($class, $uri_or_re, $app_or_res) = @_;
     $app_or_res //= [200, [], []];
-    my $old_handler = $Handlers->{$uri_or_re};
+    my $old_handler = $Intercepter->get($uri_or_re);
```
Comment (astj) [inline lib/Test/WWW/Stub.pm]: This method name seems not to be obvious to return a handler.
How about renaming this method to something like `get_handler` ?

File: lib/Test/WWW/Stub/Handler.pm [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    }
+    return $class->new(pattern => $uri_or_re, app => $app);
+}
+
+sub is_regexp {
+    my ($self) = @_;
+    return ref($self->{pattern}) eq 'Regexp';
+}
+
+sub is_str {
```
Comment (astj) [inline lib/Test/WWW/Stub/Handler.pm]: evaluating `is_str` against an instance of `URI.pm` returns true, but IMO it should return false.
Though current behavior of `_match` should be kept unchanged (We want to match an instance of `URI.pm` by string comparison), but this method name `is_str` is misleading.

```perl
use strict;
use warnings;
use Test::More;

use URI;
use Test::WWW::Stub::Handler;

my $handler = Test::WWW::Stub::Handler->new(
    pattern => URI->new('http://www.google.com/'),
    app => sub { [] },
);

ok !$handler->is_regexp;
{
    local $TODO = 'pattern is an instance of URI, so `is_str` should not be true';
    ok !$handler->is_str;
}

done_testing;

__END__
$ carton exec -- prove -lvr t/20_handler.t
t/20_handler.t ..
ok 1
not ok 2 # TODO pattern is an instance of URI, so `is_str` should not be true

#   Failed (TODO) test at t/20_handler.t line 16.
1..2
ok
All tests successful.
Files=1, Tests=2,  0 wallclock secs ( 0.02 usr  0.00 sys +  0.02 cusr  0.00 csys =  0.04 CPU)
Result: PASS
```

**Comment by astj** [comment]
@aereal Thank you for refactoring!  This looks very good, but I wrote some comments. Please check them.

File: lib/Test/WWW/Stub/Handler.pm (hunk 2 above)
Comment (aereal) [inline lib/Test/WWW/Stub/Handler.pm]: OK, all the `Handler` should know is how it compared, but not what it is.

I'll remove `is_str` and `is_regexp` and add `is_static_pattern` that means `pattern` should be compared with the URI using `eq` operator or not.
Do you think about it?

**Comment by aereal** [comment]
- [x] Add `Test::WWW::Stub::{Handler,Interceptor}` to `00_compile.t`
- [x] Rename `Test::WWW::Stub::Interceptor`
- [ ] Rename `Test::WWW::Stub::Handler#is_str`

File: lib/Test/WWW/Stub/Handler.pm (hunk 2 above)
Comment (astj) [inline lib/Test/WWW/Stub/Handler.pm]: :+1: :bow:

**Comment by aereal** [comment]
@astj I'd fixed reviewed issues!

**Comment by astj** [comment]
@aereal Thanx!! LGTM!!

**Comment by astj** [comment]
And merged!

**Comment by aereal** [comment]
:pray:

**Pull request merged**